--- a/pppd/plugins/pptp/pptp.c
+++ b/pppd/plugins/pptp/pptp.c
@@ -124,6 +124,7 @@ static int pptp_start_client(void)
 	socklen_t len;
 	struct sockaddr_pppox src_addr,dst_addr;
 	struct hostent *hostinfo;
+	u_int16_t peer_call_id = 0;
 
 	hostinfo=gethostbyname(pptp_server);
   if (!hostinfo)
@@ -183,7 +184,8 @@ static int pptp_start_client(void)
 		return -1;
         }
         /* Exchange PIDs, get call ID */
-    } while (get_call_id(callmgr_sock, getpid(), getpid(), &dst_addr.sa_addr.pptp.call_id) < 0);
+    } while (get_call_id(callmgr_sock, getpid(), getpid(), &peer_call_id) < 0);
+    dst_addr.sa_addr.pptp.call_id = peer_call_id;
 
 	if (connect(pptp_fd,(struct sockaddr*)&dst_addr,sizeof(dst_addr)))
 	{
